### PR TITLE
HexLayout

### PR DESCRIPTION
**Pull request by jsuarezruiz**
### Description of Change ###

A Layout that arranges the elements in a honeycomb pattern.

xct-hexlayout

### API Changes ###

Added: 
 
- HexLayout

```
<layout:HexLayout    
    Orientation="Vertical"
    RowCount="5" 
    ColumnCount="5">
    <BoxView layout:HexLayout.Row="0" layout:HexLayout.Column="1" CornerRadius="24" BackgroundColor="#006699" Margin="12"/>
    <BoxView layout:HexLayout.Row="0" layout:HexLayout.Column="2" CornerRadius="24" BackgroundColor="#0033CC" Margin="12"/>
    <BoxView layout:HexLayout.Row="0" layout:HexLayout.Column="3" CornerRadius="24" BackgroundColor="#3333FF" Margin="12"/>
    <BoxView layout:HexLayout.Row="1" layout:HexLayout.Column="0" CornerRadius="24" BackgroundColor="#00cc99" Margin="12"/>
    <BoxView layout:HexLayout.Row="1" layout:HexLayout.Column="1" CornerRadius="24" BackgroundColor="#00FFFF" Margin="12"/>
    <BoxView layout:HexLayout.Row="1" layout:HexLayout.Column="2" CornerRadius="24" BackgroundColor="#3399FF" Margin="12"/>
    <BoxView layout:HexLayout.Row="1" layout:HexLayout.Column="3" CornerRadius="24" BackgroundColor="#6600FF" Margin="12"/>
    <BoxView layout:HexLayout.Row="2" layout:HexLayout.Column="0" CornerRadius="24" BackgroundColor="#008000" Margin="12"/>
    <BoxView layout:HexLayout.Row="2" layout:HexLayout.Column="1" CornerRadius="24" BackgroundColor="#66FF66" Margin="12"/>
    <BoxView layout:HexLayout.Row="2" layout:HexLayout.Column="2" CornerRadius="24" BackgroundColor="#FFFFFF" Margin="12"/>
    <BoxView layout:HexLayout.Row="2" layout:HexLayout.Column="3" CornerRadius="24" BackgroundColor="#FF66FF" Margin="12"/>
    <BoxView layout:HexLayout.Row="2" layout:HexLayout.Column="4" CornerRadius="24" BackgroundColor="#CC00CC" Margin="12"/>
    <BoxView layout:HexLayout.Row="3" layout:HexLayout.Column="0" CornerRadius="24" BackgroundColor="#99CC00" Margin="12"/>
    <BoxView layout:HexLayout.Row="3" layout:HexLayout.Column="1" CornerRadius="24" BackgroundColor="#FFFF66" Margin="12"/>
    <BoxView layout:HexLayout.Row="3" layout:HexLayout.Column="2" CornerRadius="24" BackgroundColor="#FF7C80" Margin="12"/>
    <BoxView layout:HexLayout.Row="3" layout:HexLayout.Column="3" CornerRadius="24" BackgroundColor="#D60093" Margin="12"/>
    <BoxView layout:HexLayout.Row="4" layout:HexLayout.Column="1" CornerRadius="24" BackgroundColor="#CC9900" Margin="12"/>
    <BoxView layout:HexLayout.Row="4" layout:HexLayout.Column="2" CornerRadius="24" BackgroundColor="#FF3300" Margin="12"/>
    <BoxView layout:HexLayout.Row="4" layout:HexLayout.Column="3" CornerRadius="24" BackgroundColor="#CC0000" Margin="12"/>
</layout:HexLayout>
```

### PR Checklist ###

- [ ] Has tests (if omitted, state reason in description)
- [x] Has samples (if omitted, state reason in description)
- [ ] Rebased on top of main at time of PR
- [ ] Changes adhere to coding standard

### CODE REVIEW

File: XamarinCommunityToolkit/Layouts/HexLayout.shared.cs [hunk 1]
```diff
@@ -0,0 +1,278 @@
+﻿using System;
+using Xamarin.Forms;
+
+namespace Xamarin.CommunityToolkit.Layouts
```
Comment (jfversluis) [inline XamarinCommunityToolkit/Layouts/HexLayout.shared.cs:4]: There is a test failing over this one. I see we technically have in the README that Layouts is another namespace (but then .UI.Layouts) do we want to actually have that or put it under Views?

Either way is fine with me but then we need to update the test :)

Comment (AndreiMisiukevich) [reply XamarinCommunityToolkit/Layouts/HexLayout.shared.cs:4]: I'd prefer to put it inside Views

Actually, Expander and SideMenu are a layouts too

Comment (jsuarezruiz) [reply XamarinCommunityToolkit/Layouts/HexLayout.shared.cs:4]: Ok, I am going to move it to Views namespace (and folder).

**Comment by jfversluis** [comment]
A new build for this PR has been completed. As a result a NuGet has been created which allows you to verify this fix or try out the new feature! [Download here](https://dev.azure.com/xamarin/6fd3d886-57a5-4e31-8db7-52a1b47c07a8/_apis/build/builds/29746/artifacts?artifactName=nuget&api-version=6.1-preview.5&%24format=zip)

File: XamarinCommunityToolkit/Layouts/HexLayout.shared.cs (hunk 1 above)
Comment (jsuarezruiz) [reply XamarinCommunityToolkit/Layouts/HexLayout.shared.cs]: Moved.

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/HexLayout.shared.cs [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+    /// A Layout that arranges the elements in a honeycomb pattern.
+    /// Based on https://github.com/AlexanderSharykin/HexGrid. Created by Alexander Sharykin.
+    /// </summary>
+    public class HexLayout : Layout<View>
+    {
+        public static readonly BindableProperty OrientationProperty =
+         BindableProperty.Create(nameof(Orientation), typeof(StackOrientation), typeof(HexLayout), StackOrientation.Vertical,
+             BindingMode.TwoWay, propertyChanged: (bindable, oldvalue, newvalue) => ((HexLayout)bindable).InvalidateMeasure());
+
+        public StackOrientation Orientation
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/HexLayout.shared.cs:16]: I wonder whether we should expose our own enum HexOrientation?

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/HexLayout.shared.cs [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+            set => SetValue(ColumnProperty, value);
+        }
+
+        int GetColumn(VisualElement e)
+        {
+            var column = (int)e.GetValue(ColumnProperty);
+            if (column >= ColumnCount)
+                column = ColumnCount - 1;
+            return column;
+        }
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/HexLayout.shared.cs:58]:
```csharp
return Math.Min((int)e.GetValue(ColumnProperty), ColumnCount - 1);
```

Not sure that's my solution is better 🤓

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/HexLayout.shared.cs (hunk 2 above)
Comment (jsuarezruiz) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Views/HexLayout.shared.cs:16]: Yes, added `HexOrientation`.

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/HexLayout.shared.cs (hunk 3 above)
Comment (jsuarezruiz) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Views/HexLayout.shared.cs:58]: Mnn, I think is easier to read. Changed.

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/HexLayout/HexLayout.shared.cs [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+
+        public int RowCount
+        {
+            get => (int)GetValue(RowCountProperty);
+            set => SetValue(RowCountProperty, value);
+        }
+
+        public static readonly BindableProperty ColumnProperty =
+            BindableProperty.Create(nameof(Column), typeof(int), typeof(HexLayout), 1,
+                BindingMode.TwoWay, null);
```
Comment (pictos) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/HexLayout/HexLayout.shared.cs:44]: Can we remove this `null`?
```suggestion
        public static readonly BindableProperty ColumnProperty =
            BindableProperty.Create(nameof(Column), typeof(int), typeof(HexLayout), 1,
                BindingMode.TwoWay);
```

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/HexLayout/HexLayout.shared.cs [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+        {
+            get => (int)GetValue(ColumnProperty);
+            set => SetValue(ColumnProperty, value);
+        }
+
+        int GetColumn(VisualElement e) => Math.Min((int)e.GetValue(ColumnProperty), ColumnCount - 1);
+
+        public static readonly BindableProperty RowProperty =
+            BindableProperty.Create(nameof(Row), typeof(int), typeof(HexLayout), 1,
+                BindingMode.TwoWay, null);
```
Comment (pictos) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/HexLayout/HexLayout.shared.cs:56]: Can you remove this `null`?
```suggestion
        public static readonly BindableProperty RowProperty =
            BindableProperty.Create(nameof(Row), typeof(int), typeof(HexLayout), 1,
                BindingMode.TwoWay);
```

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/HexLayout/HexLayout.shared.cs [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+{
+    /// <summary>
+    /// A Layout that arranges the elements in a honeycomb pattern.
+    /// Based on https://github.com/AlexanderSharykin/HexGrid. Created by Alexander Sharykin.
+    /// </summary>
+    public class HexLayout : Layout<View>
+    {
+        public static readonly BindableProperty OrientationProperty =
+         BindableProperty.Create(nameof(Orientation), typeof(HexOrientation), typeof(HexLayout), HexOrientation.Vertical,
+             BindingMode.TwoWay, propertyChanged: (bindable, oldvalue, newvalue) => ((HexLayout)bindable).InvalidateMeasure());
```
Comment (pictos) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/HexLayout/HexLayout.shared.cs:14]: Can we refactor that property changed to a method?

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/HexLayout/HexLayout.shared.cs [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+
+        public HexOrientation Orientation
+        {
+            get => (HexOrientation)GetValue(OrientationProperty);
+            set => SetValue(OrientationProperty, value);
+        }
+
+        public static readonly BindableProperty ColumnCountProperty =
+            BindableProperty.Create(nameof(ColumnCount), typeof(int), typeof(HexLayout), 1,
+                BindingMode.TwoWay, propertyChanged: (bindable, oldvalue, newvalue) => ((HexLayout)bindable).InvalidateMeasure());
```
Comment (pictos) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/HexLayout/HexLayout.shared.cs:24]: Can we refactor that property changed to a method?

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/HexLayout/HexLayout.shared.cs [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+
+                foreach (var child in Children)
+                {
+                    var childSize = child.Measure(widthConstraint, heightConstraint);
+
+                    if (childSize.Request.Height > h)
+                        h = childSize.Request.Height;
+
+                    if (childSize.Request.Width > w)
+                        w = childSize.Request.Width;
```
Comment (pictos) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/HexLayout/HexLayout.shared.cs:84]:
```suggestion
					h = Math.Max(h, childSize.Request.Height);
					w = Math.Max(w, childSize.Request.Width);
```

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/HexLayout/HexLayout.shared.cs [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+
+                return new SizeRequest(new Size(w * ((ColumnCount * 2) + 1) / 2, h * ((RowCount * 3) + 1) / 4));
+            }
+
+            return Measure(widthConstraint, heightConstraint);
+        }
+
+        protected override void LayoutChildren(double x, double y, double width, double height)
+        {
+            HasShift(out var first, out var last);
```
Comment (pictos) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/HexLayout/HexLayout.shared.cs:98]: We aren't using the `last` value.
```suggestion
            HasShift(out var first, out _);
```

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/HexLayout/HexLayout.shared.cs [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+                columnWidth = 0.25 * hexSize.Width;
+            }
+            else
+            {
+                rowHeight = 0.25 * hexSize.Height;
+                columnWidth = 0.5 * hexSize.Width;
+            }
+
+            var elements = Children;
+            for (var i = 0; i < elements.Count; i++)
```
Comment (pictos) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/HexLayout/HexLayout.shared.cs:116]:
```suggestion
            var elementsCount = elements.Count;
            for (var i = 0; i < elementsCount; i++)
```

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/HexLayout/HexLayout.shared.cs [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+            else
+                HasColumnShift(out first, out last);
+        }
+
+        void HasRowShift(out bool firstRow, out bool lastRow)
+        {
+            firstRow = lastRow = true;
+
+            var elements = Children;
+            for (var i = 0; i < elements.Count && (firstRow || lastRow); i++)
```
Comment (pictos) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/HexLayout/HexLayout.shared.cs:160]:
```suggestion
            var elementsCount = elements.Count;
            for (var i = 0; i < elements.Count && (firstRow || lastRow); i++)
```

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/HexLayout/HexLayout.shared.cs [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+                    lastRow = false;
+            }
+        }
+
+        void HasColumnShift(out bool firstColumn, out bool lastColumn)
+        {
+            firstColumn = lastColumn = true;
+
+            var elements = Children;
+            for (var i = 0; i < elements.Count && (firstColumn || lastColumn); i++)
```
Comment (pictos) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/HexLayout/HexLayout.shared.cs:185]:
```suggestion
            var elementsCount = elements.Count;
            for (var i = 0; i < elementsCount && (firstColumn || lastColumn); i++)
```

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/HexLayout/HexLayout.shared.cs [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+
+            foreach (var e in Children)
+            {
+                if (e is VisualElement f)
+                {
+                    if (f.MinimumHeightRequest > minH)
+                        minH = f.MinimumHeightRequest;
+
+                    if (f.MinimumWidthRequest > minW)
+                        minW = f.MinimumWidthRequest;
```
Comment (pictos) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/HexLayout/HexLayout.shared.cs:218]:
```suggestion
					minH = Math.Max(minH, f.MinimumHeightRequest);
					minW = Math.Max(minW, f.MinimumWidthRequest);
```

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/HexLayout/HexLayout.shared.cs [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+                        h = childSize.Request.Height;
+
+                    if (childSize.Request.Width > w)
+                        w = childSize.Request.Width;
+                }
+
+                if (Orientation == HexOrientation.Horizontal)
+                    return new SizeRequest(new Size(w * ((ColumnCount * 3) + 1) / 4, h * ((RowCount * 2) + 1) / 2));
+
+                return new SizeRequest(new Size(w * ((ColumnCount * 2) + 1) / 2, h * ((RowCount * 3) + 1) / 4));
```
Comment (pictos) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/HexLayout/HexLayout.shared.cs:90]: Wondering if we should cache these values. I'm not an expert on layouts, but from the [docs](https://docs.microsoft.com/en-us/xamarin/xamarin-forms/user-interface/layouts/custom#calculate-and-cache-layout-data), it's a good practice to do it, but I don't know if it fits in this case.

**Review by pictos — CHANGES_REQUESTED** [review]
Thanks for this @jsuarezruiz! I ran here and works as expected. My suggestions are more about code style and a couple of perf improviments.

**Comment by jfversluis** [comment]
Can we retarget this one to develop so it ends up in 1.1?

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/HexLayout/HexLayout.shared.cs (hunk 6 above)
Comment (jsuarezruiz) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Views/HexLayout/HexLayout.shared.cs:14]: Changed.

**Review by AndreiMisiukevich — APPROVED** [review]

**Comment by AndreiMisiukevich** [comment]
@pictos please review/approve (can't merge cuz changes were requested from your side)

**Review by pictos — APPROVED** [review]

**Pull request merged**